extendr-api/src/robj/try_from_robj.rs

```diff
@@ -102,6 +103,123 @@ impl_try_from_scalar_integer!(isize);
+                            Ok(value)
+                        }
+                    }
+                    _ => Err(Error::ExpectedScalar(robj.clone())),
+                }
+            }
+        }
+
+        impl<'a> TryFrom<&'a mut Robj> for &'a mut $type
+        where
+            Robj: AsTypedSlice<'a, $type>,
+        {
+            type Error = Error;
+
+            fn try_from(robj: &'a mut Robj) -> Result<Self> {
+                let len = robj.len();
+                if len == 0 {
+                    return Err(Error::ExpectedNonZeroLength(robj.clone()));
+                }
+                if len != 1 {
+                    return Err(Error::ExpectedScalar(robj.clone()));
+                }
+
+                if robj.as_typed_slice().is_none() {
+                    return Err(Error::$err_variant(robj.clone()));
+                }
+
+                let slice: &mut [$type] = match robj.as_typed_slice_mut() {
+                    Some(s) => s,
+                    None => return Err(Error::$err_variant(robj.clone())),
+                };
+                let value = &mut slice[0];
+
+                if !$allow_na && $is_na(value) {
+                    return Err(Error::MustNotBeNA(robj.clone()));
+                }
+
+                Ok(value)
+            }
+        }
+
+        impl<'a> TryFrom<&'a Robj> for Option<&'a $type>
+        where
+            Robj: AsTypedSlice<'a, $type>,
+        {
+            type Error = Error;
+
+            fn try_from(robj: &'a Robj) -> Result<Self> {
+                if robj.is_null() || robj.is_na() {
+                    Ok(None)
+                } else {
+                    <&$type>::try_from(robj).map(Some)
+                }
+            }
+        }
+
+        impl<'a> TryFrom<&'a mut Robj> for Option<&'a mut $type>
+        where
+            Robj: AsTypedSlice<'a, $type>,
+        {
+            type Error = Error;
+
+            fn try_from(robj: &'a mut Robj) -> Result<Self> {
+                if robj.is_null() || robj.is_na() {
+                    Ok(None)
+                } else {
+                    <&mut $type>::try_from(robj).map(Some)
+                }
+            }
+        }
+    };
+}
+
+impl_scalar_borrow!(i32, ExpectedInteger, false, |v: &i32| v.is_na());
+impl_scalar_borrow!(f64, ExpectedReal, false, |v: &f64| v.is_na() || v.is_nan());
+impl_scalar_borrow!(Rint, ExpectedInteger, true, |v: &Rint| v.is_na());
+impl_scalar_borrow!(Rfloat, ExpectedReal, true, |v: &Rfloat| v.is_na());
+impl_scalar_borrow!(Rbool, ExpectedLogical, true, |v: &Rbool| v.is_na());
+impl_scalar_borrow!(Rcplx, ExpectedComplex, true, |v: &Rcplx| {
+    let re = v.re().inner();
+    let im = v.im().inner();
+    re.is_nan() || im.is_nan() || re.is_na() || im.is_na()
+});
+impl_scalar_borrow!(c64, ExpectedComplex, true, |v: &c64| {
+    let rc: Rcplx = (*v).into();
+    let re = rc.re().inner();
+    let im = rc.im().inner();
+    re.is_nan() || im.is_nan() || re.is_na() || im.is_na()
+});
```

Comment:
should `impl_scalar_borrow!(c64, ...` be false for allow_na? It is a rust struct built from two `f64` values and `impl_scalar_borrow!` for `f64` is false.

Reply:
Good catch!